File: app/src/main/java/com/robotnec/budget/core/mvp/presenter/AddAccountPresenter.java
package com.robotnec.budget.core.mvp.presenter;

import com.robotnec.budget.core.persistence.dao.AccountDao;
import com.robotnec.budget.core.di.ApplicationComponent;
import com.robotnec.budget.core.domain.Account;
import com.robotnec.budget.core.domain.Currency;
import com.robotnec.budget.core.domain.MoneyAmount;
import com.robotnec.budget.core.mvp.view.AddAccountView;

import javax.inject.Inject;

/**
 * @author zak <zak@swingpulse.com>
 */
public class AddAccountPresenter extends Presenter<AddAccountView> {

    @Inject
    AccountDao accountDao;

    private final Account account;
    private final boolean editMode;

    public AddAccountPresenter(AddAccountView view, Account account) {
        super(view);
        this.account = account;
        editMode = this.account != null;
    }

    @Override
    public void injectComponent(ApplicationComponent applicationComponent) {
        applicationComponent.inject(this);
    }

    @Override
    public void onViewResume() {
        if (editMode) {
            view.initEditMode(account);
        }
    }

    public void addOrEditAccount(String accountName, String accountAmount, String accountCurrency) {
        MoneyAmount amount = MoneyAmount.of(Double.parseDouble(accountAmount),
                Currency.fromCode(accountCurrency));
        if (editMode) {
            account.setName(accountName);
            account.setAmount(amount);
            accountDao.createOrUpdate(account);
        } else {
            Account account = new Account();
            account.setName(accountName);
            account.setAmount(amount);
            accountDao.createOrUpdate(account);
        }
    }

    public void deleteAccount() {
        accountDao.remove(account.getId());
        view.closeView();
    }
}

File: app/src/main/java/com/robotnec/budget/core/mvp/presenter/AddAccountPresenter.kt
package com.robotnec.budget.core.mvp.presenter

import com.robotnec.budget.refactored.di.ApplicationComponent
import com.robotnec.budget.core.domain.Account
import com.robotnec.budget.core.mvp.view.AddAccountView
import com.robotnec.budget.core.persistence.dao.AccountDao
import org.joda.money.CurrencyUnit
import org.joda.money.Money
import javax.inject.Inject

/**
 * @author zak zak@swingpulse.com>
 */
class AddAccountPresenter(view: AddAccountView, private val account: Account?) : Presenter<AddAccountView>(view) {

    @Inject
    internal lateinit var accountDao: AccountDao
    private val editMode: Boolean = this.account != null

    override fun injectComponent(applicationComponent: ApplicationComponent) {
        applicationComponent.inject(this)
    }

    override fun onViewResume() {
        if (editMode) {
            view.initEditMode(account!!)
        }
    }

    fun addOrEditAccount(accountName: String, accountAmount: String, accountCurrency: String) {
        val amount = Money.of(CurrencyUnit.of(accountCurrency), accountAmount.toDouble())
        if (editMode) {
            account!!.name = accountName
            account.amount = amount
            accountDao.createOrUpdate(account)
        } else {
            val account = Account(-1, accountName, amount)
            accountDao.createOrUpdate(account)
        }
    }

    fun deleteAccount() {
        accountDao.remove(account!!.id)
        view.closeView()
    }
}
